macros/src/lib.rs

```diff
@@ -8,194 +8,59 @@
-                is_unsafe: false,
-                unify_fieldless_variants: false,
-                generics: LifetimeBounds::empty(),
-                explicit_self: None,
-                args: vec!(Literal(path_local!(i64))),
-                ret_ty: Literal(Path::new_(pathvec_std!(cx, core::option::Option),
-                                           None,
-                                           vec!(Box::new(Self_)),
-                                           true)),
-                // #[inline] liable to cause code-bloat
-                attributes: attrs.clone(),
-                combine_substructure: combine_substructure(Box::new(|c, s, sub| {
-                    cs_from("i64", c, s, sub)
-                })),
-            },
-            MethodDef {
-                name: "from_u64",
-                is_unsafe: false,
-                unify_fieldless_variants: false,
-                generics: LifetimeBounds::empty(),
-                explicit_self: None,
-                args: vec!(Literal(path_local!(u64))),
-                ret_ty: Literal(Path::new_(pathvec_std!(cx, core::option::Option),
-                                           None,
-                                           vec!(Box::new(Self_)),
-                                           true)),
-                // #[inline] liable to cause code-bloat
-                attributes: attrs,
-                combine_substructure: combine_substructure(Box::new(|c, s, sub| {
-                    cs_from("u64", c, s, sub)
-                })),
-            }
-        ),
-        associated_types: Vec::new(),
-        supports_unions: false,
+    let variants = match ast.body {
+        Enum(ref variants) => variants,
+        _ => panic!("`FromPrimitive` can be applied only to the enums, {} is not an enum", name),
     };
 
-    trait_def.expand(cx, mitem, &item, push)
-}
-
-fn cs_from(name: &str, cx: &mut ExtCtxt, trait_span: Span, substr: &Substructure) -> P<Expr> {
-    if substr.nonself_args.len() != 1 {
-        cx.span_bug(trait_span, "incorrect number of arguments in `derive(FromPrimitive)`")
-    }
-
-    let n = &substr.nonself_args[0];
-
-    match *substr.fields {
-        StaticStruct(..) => {
-            cx.span_err(trait_span, "`FromPrimitive` cannot be derived for structs");
-            return cx.expr_fail(trait_span, InternedString::new(""));
-        }
-        StaticEnum(enum_def, _) => {
-            if enum_def.variants.is_empty() {
-                cx.span_err(trait_span,
-                            "`FromPrimitive` cannot be derived for enums with no variants");
-                return cx.expr_fail(trait_span, InternedString::new(""));
+    let mut idx = 0;
+    let variants: Vec<_> = variants.iter()
+        .map(|variant| {
+            let ident = &variant.ident;
+            if let Some(val) = variant.discriminant {
+                idx = val.value;
+            }
+            let tt = quote!(#idx => Some(#name::#ident));
```

Comment:
Remember that enums can have data too, but `FromPrimitive` only makes sense for integral enums.  It looks like that's where `Variant.data` is `VariantData::Unit`.  If any variant is something else, we need to reject the whole thing.  

For instance, try adding something like `Other(u8, u8, u8)` to one of the `Color` tests.  It does get an error, but it's a very confusing one:

```
error[E0308]: mismatched types
  --> tests/trivial.rs:18:1
   |
18 | enum Color {
   | ^ expected enum `Color`, found fn item
   |
   = note: expected type `Color`
   = note:    found type `fn(u8, u8, u8) -> Color {Color::Other}`
```

The derivation tried to output `Some(Color::Other)` here, but that's actually a type constructor, not a value in itself, which is why the error is talking about a `fn`.  Similarly if it were `Other{ r:u8, g:u8, b:u8 }` you get the error `struct called like a function`.  We should catch this up front to identify the root problem.

We could use [compiletest-rs](https://github.com/laumann/compiletest-rs) to have some negative tests.